`resource_missing_tags` panic: value is unknown

In [v0.24.2](https://github.com/terraform-linters/tflint-ruleset-aws/releases/tag/v0.24.2) the following error occurs:

```
% tflint
Failed to check ruleset; error reading from server: EOF
```
The TFlint version used is v0.47.0.

Output with TFLINT_LOG=debug tflint:
```
discovery.go:54: [INFO] Plugin `aws` found
go-plugin@v1.4.10/client.go:604: starting plugin: path=../../../../../../.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.24.2/tflint-ruleset-aws args=["../../../../../../.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.24.2/tflint-ruleset-aws"]
go-plugin@v1.4.10/client.go:612: plugin started: path=../../../../../../.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.24.2/tflint-ruleset-aws pid=1537
go-plugin@v1.4.10/client.go:707: waiting for RPC address: path=../../../../../../.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.24.2/tflint-ruleset-aws
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: go-plugin@v1.4.10/server.go:404: plugin address: network=unix address=/tmp/plugin2985356507
go-plugin@v1.4.10/client.go:752: using plugin: version=11
discovery.go:33: [INFO] Plugin `terraform` is not installed, but the bundled plugin is available.
discovery.go:54: [INFO] Plugin `terraform` found
go-plugin@v1.4.10/client.go:604: starting plugin: path=/usr/bin/tflint args=["/usr/bin/tflint", "--act-as-bundled-plugin"]
go-plugin@v1.4.10/client.go:612: plugin started: path=/usr/bin/tflint pid=1547
go-plugin@v1.4.10/client.go:707: waiting for RPC address: path=/usr/bin/tflint
go-plugin@v1.4.10/client.go:752: using plugin: version=11
go-plugin@v1.4.10/client.go:1046: tflint: go-plugin@v1.4.10/server.go:404: plugin address: network=unix address=/tmp/plugin42334417
host2plugin/client.go:124: starting host-side gRPC server
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: plugin2host/client.go:398: unknown value found in ods2/.terraform/modules/lambda-event-source-mapping-ods2-fault-tolerance-critical-path/modules/aws/lambda/eventSQS/main.tf:2,40-60
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: plugin2host/client.go:398: unknown value found in ods2/.terraform/modules/lambda-event-source-mapping-ods2-fault-tolerance-critical-path/modules/aws/lambda/eventSQS/main.tf:3,40-57
host2plugin/client.go:124: starting host-side gRPC server
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: rules/aws_resource_missing_tags.go:204: Walk `%s` attribute: EXTRA_VALUE_AT_END=aws_sns_topic.main.tags
host2plugin/client.go:124: starting host-side gRPC server
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: rules/aws_resource_missing_tags.go:204: Walk `%s` attribute: EXTRA_VALUE_AT_END=aws_s3_bucket.main.tags
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: panic: value is unknown
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: goroutine 54 [running]:
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/zclconf/go-cty/cty.Value.AsString({{{0x1f9f868?, 0xc0001272b1?}}, {0x17b8b60?, 0xc0005f5220?}})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/zclconf/go-cty@v1.13.2/cty/value_ops.go:1393 +0x111
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-ruleset-aws/rules.(*AwsResourceMissingTagsRule).Check.func1({{{0x1f9f948?, 0xc0005f5440?}}, {0x17caaa0?, 0xc000e05830?}})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/work/tflint-ruleset-aws/tflint-ruleset-aws/rules/aws_resource_missing_tags.go:221 +0x44b
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: reflect.Value.call({0x16ef2e0?, 0xc000cb7380?, 0x10?}, {0x1bb6859, 0x4}, {0xc000edd580, 0x1, 0xa?})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/opt/hostedtoolcache/go/1.20.5/x64/src/reflect/value.go:586 +0xb0b
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: reflect.Value.Call({0x16ef2e0?, 0xc000cb7380?, 0xc00055e1e0?}, {0xc000edd580?, 0xc000ca4b20?, 0x1?})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/opt/hostedtoolcache/go/1.20.5/x64/src/reflect/value.go:370 +0xbc
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-plugin-sdk/plugin/plugin2host.(*GRPCClient).EvaluateExpr(0x0?, {0x1f9f130, 0xc00055e1e0}, {0x16ef2e0, 0xc000cb7380?}, 0x1?)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/terraform-linters/tflint-plugin-sdk@v0.17.0/plugin/plugin2host/client.go:323 +0x376
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-ruleset-aws/rules.(*AwsResourceMissingTagsRule).Check(0x2e74b78, {0x1fc1f80, 0xc00139e060})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/work/tflint-ruleset-aws/tflint-ruleset-aws/rules/aws_resource_missing_tags.go:209 +0x94b
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-plugin-sdk/plugin/host2plugin.(*GRPCServer).Check(0xc000aa36e0, {0x1f9ef00, 0xc0013d0030}, 0xc0013c8ae0?)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/terraform-linters/tflint-plugin-sdk@v0.17.0/plugin/host2plugin/server.go:145 +0x46e
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-plugin-sdk/plugin/proto._RuleSet_Check_Handler.func1({0x1f9ef00, 0xc0013d0030}, {0x192f040?, 0xc0013d0060})
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/terraform-linters/tflint-plugin-sdk@v0.17.0/plugin/proto/tflint_grpc.pb.go:354 +0x78
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-plugin-sdk/plugin/interceptor.RequestLogging.func1({0x1f9ef00, 0xc0013d0030}, {0x192f040?, 0xc0013d0060?}, 0xc00139fce0, 0xc0013c2d20)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/terraform-linters/tflint-plugin-sdk@v0.17.0/plugin/interceptor/logging.go:16 +0x20f
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: github.com/terraform-linters/tflint-plugin-sdk/plugin/proto._RuleSet_Check_Handler({0x1ac92e0?, 0xc000aa36e0}, {0x1f9ef00, 0xc0013d0030}, 0xc000e18000, 0xc00012e300)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/github.com/terraform-linters/tflint-plugin-sdk@v0.17.0/plugin/proto/tflint_grpc.pb.go:356 +0x138
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: google.golang.org/grpc.(*Server).processUnaryRPC(0xc00067e3c0, {0x1fbf660, 0xc000502820}, 0xc0013c7320, 0xc000aa3770, 0x2e2f880, 0x0)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:1337 +0xdf3
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: google.golang.org/grpc.(*Server).handleStream(0xc00067e3c0, {0x1fbf660, 0xc000502820}, 0xc0013c7320, 0x0)
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:1714 +0xa36
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: google.golang.org/grpc.(*Server).serveStreams.func1.1()
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:959 +0x98
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: created by google.golang.org/grpc.(*Server).serveStreams.func1
go-plugin@v1.4.10/client.go:1046: tflint-ruleset-aws: 	/home/runner/go/pkg/mod/google.golang.org/grpc@v1.55.0/server.go:957 +0x18c
go-plugin@v1.4.10/grpc_stdio.go:142: stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
19:31:00 [ERROR] go-plugin@v1.4.10/client.go:662: plugin process exited: path=../../../../../../.tflint.d/plugins/github.com/terraform-linters/tflint-ruleset-aws/0.24.2/tflint-ruleset-aws pid=1537 error="exit status 2"
go-plugin@v1.4.10/client.go:472: plugin exited
go-plugin@v1.4.10/grpc_stdio.go:142: stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
go-plugin@v1.4.10/client.go:665: plugin process exited: path=/usr/bin/tflint pid=1547
go-plugin@v1.4.10/client.go:472: plugin exited
Failed to check ruleset; error reading from server: EOF
```

The `.tflint.hcl`:
```
config {
  plugin_dir = "../../../../../../.tflint.d/plugins"

  module              = true
  force               = false
  disabled_by_default = false
}

plugin "aws" {
  enabled = true
  version = "0.24.1"
  source  = "github.com/terraform-linters/tflint-ruleset-aws"
}

rule "aws_resource_missing_tags" {
  enabled = true
  tags    = ["Name", "Application"]
}

rule "terraform_typed_variables" {
  enabled = false
}

rule "terraform_unused_declarations" {
  enabled = false
}
```

--- a/rules/aws_resource_missing_tags.go
+++ b/rules/aws_resource_missing_tags.go
@@ -14,7 +14,7 @@ import (
 	"github.com/terraform-linters/tflint-ruleset-aws/project"
 	"github.com/terraform-linters/tflint-ruleset-aws/rules/tags"
 	"github.com/zclconf/go-cty/cty"
-	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
 )
 
 // AwsResourceMissingTagsRule checks whether resources are tagged correctly
@@ -59,7 +59,7 @@ func (r *AwsResourceMissingTagsRule) Link() string {
 	return project.ReferenceLink(r.Name())
 }
 
-func (r *AwsResourceMissingTagsRule) getProviderLevelTags(runner tflint.Runner) (map[string]map[string]string, error) {
+func (r *AwsResourceMissingTagsRule) getProviderLevelTags(runner tflint.Runner) (map[string][]string, error) {
 	providerSchema := &hclext.BodySchema{
 		Attributes: []hclext.AttributeSchema{
 			{
@@ -81,7 +81,7 @@ func (r *AwsResourceMissingTagsRule) getProviderLevelTags(runner tflint.Runner) 
 	}
 
 	// Get provider default tags
-	allProviderTags := make(map[string]map[string]string)
+	allProviderTags := make(map[string][]string)
 	var providerAlias string
 	for _, provider := range providerBody.Blocks.OfType(providerAttributeName) {
 		// Get the alias attribute, in terraform when there is a single aws provider its called "default"
@@ -102,29 +102,22 @@ func (r *AwsResourceMissingTagsRule) getProviderLevelTags(runner tflint.Runner) 
 		}
 
 		for _, block := range provider.Body.Blocks {
-			providerTags := make(map[string]string)
+			var providerTags []string
 			attr, ok := block.Body.Attributes[tagsAttributeName]
 			if !ok {
 				continue
 			}
 
 			err := runner.EvaluateExpr(attr.Expr, func(val cty.Value) error {
-				// Skip unknown values
-				if !val.IsKnown() || val.IsNull() {
+				keys, known := getKeysForValue(val)
+
+				if !known {
 					logger.Warn("The missing aws tags rule can only evaluate provided variables, skipping %s.", provider.Labels[0]+"."+providerAlias+"."+defaultTagsBlockName+"."+tagsAttributeName)
 					return nil
 				}
 
-				valMap := val.AsValueMap()
-				var tags map[string]string = make(map[string]string, len(valMap))
-				for k, v := range valMap {
-					tags[k] = ""
-					if !v.IsNull() && v.Type().FriendlyName() == "string" {
-						tags[k] = v.AsString()
-					}
-				}
-				logger.Debug("Walk `%s` provider with tags `%v`", providerAlias, tags)
-				providerTags = tags
+				logger.Debug("Walk `%s` provider with tags `%v`", providerAlias, keys)
+				providerTags = keys
 				return nil
 			}, nil)
 
@@ -193,12 +186,6 @@ func (r *AwsResourceMissingTagsRule) Check(runner tflint.Runner) error {
 				}
 			}
 
-			resourceTags := make(map[string]string)
-
-			// The provider tags are to be overriden
-			// https://registry.terraform.io/providers/hashicorp/aws/latest/docs#default_tags
-			maps.Copy(resourceTags, providerTagsMap[providerAlias])
-
 			// If the resource has a tags attribute
 			if attribute, okResource := resource.Body.Attributes[tagsAttributeName]; okResource {
 				logger.Debug(
@@ -207,22 +194,13 @@ func (r *AwsResourceMissingTagsRule) Check(runner tflint.Runner) error {
 				)
 
 				err := runner.EvaluateExpr(attribute.Expr, func(val cty.Value) error {
-					// Skip unknown values
-					if !val.IsKnown() || val.IsNull() {
+					keys, known := getKeysForValue(val)
+					if !known {
 						logger.Warn("The missing aws tags rule can only evaluate provided variables, skipping %s.", resource.Labels[0]+"."+resource.Labels[1]+"."+tagsAttributeName)
 						return nil
 					}
-					valMap := val.AsValueMap()
-					// Since the evlaluateExpr, overrides k/v pairs, we need to re-copy the tags
-					var evaluatedTags map[string]string = make(map[string]string, len(valMap))
-					for k, v := range valMap {
-						evaluatedTags[k] = ""
-						if !v.IsNull() && v.Type().FriendlyName() == "string" {
-							evaluatedTags[k] = v.AsString()
-						}
-					}
-					maps.Copy(resourceTags, evaluatedTags)
-					r.emitIssue(runner, resourceTags, config, attribute.Expr.Range())
+
+					r.emitIssue(runner, append(providerTagsMap[providerAlias], keys...), config, attribute.Expr.Range())
 					return nil
 				}, nil)
 
@@ -231,7 +209,7 @@ func (r *AwsResourceMissingTagsRule) Check(runner tflint.Runner) error {
 				}
 			} else {
 				logger.Debug("Walk `%s` resource", resource.Labels[0]+"."+resource.Labels[1])
-				r.emitIssue(runner, resourceTags, config, resource.DefRange)
+				r.emitIssue(runner, providerTagsMap[providerAlias], config, resource.DefRange)
 			}
 		}
 	}
@@ -283,7 +261,7 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroups(runner tflint.Run
 		case len(asgTagBlockTags) > 0 && len(asgTagsAttributeTags) > 0:
 			runner.EmitIssue(r, "Only tag block or tags attribute may be present, but found both", resource.DefRange)
 		case len(asgTagBlockTags) == 0 && len(asgTagsAttributeTags) == 0:
-			r.emitIssue(runner, map[string]string{}, config, resource.DefRange)
+			r.emitIssue(runner, []string{}, config, resource.DefRange)
 		case len(asgTagBlockTags) > 0 && len(asgTagsAttributeTags) == 0:
 			tags := asgTagBlockTags
 			location := tagBlockLocation
@@ -299,8 +277,8 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroups(runner tflint.Run
 }
 
 // checkAwsAutoScalingGroupsTag checks tag{} blocks on aws_autoscaling_group resources
-func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTag(runner tflint.Runner, config awsResourceTagsRuleConfig, resourceBlock *hclext.Block) (map[string]string, hcl.Range, error) {
-	tags := make(map[string]string)
+func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTag(runner tflint.Runner, config awsResourceTagsRuleConfig, resourceBlock *hclext.Block) ([]string, hcl.Range, error) {
+	tags := make([]string, 0)
 
 	resources, err := runner.GetResourceContent("aws_autoscaling_group", &hclext.BodySchema{
 		Blocks: []hclext.BlockSchema{
@@ -330,7 +308,7 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTag(runner tflint.
 			}
 
 			err := runner.EvaluateExpr(attribute.Expr, func(key string) error {
-				tags[key] = ""
+				tags = append(tags, key)
 				return nil
 			}, nil)
 			if err != nil {
@@ -343,8 +321,8 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTag(runner tflint.
 }
 
 // checkAwsAutoScalingGroupsTag checks the tags attribute on aws_autoscaling_group resources
-func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTags(runner tflint.Runner, config awsResourceTagsRuleConfig, resourceBlock *hclext.Block) (map[string]string, hcl.Range, error) {
-	tags := make(map[string]string)
+func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTags(runner tflint.Runner, config awsResourceTagsRuleConfig, resourceBlock *hclext.Block) ([]string, hcl.Range, error) {
+	tags := make([]string, 0)
 
 	resources, err := runner.GetResourceContent("aws_autoscaling_group", &hclext.BodySchema{
 		Attributes: []hclext.AttributeSchema{
@@ -369,7 +347,7 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTags(runner tflint
 			}))
 			err := runner.EvaluateExpr(attribute.Expr, func(asgTags []awsAutoscalingGroupTag) error {
 				for _, tag := range asgTags {
-					tags[tag.Key] = tag.Value
+					tags = append(tags, tag.Key)
 				}
 				return nil
 			}, &tflint.EvaluateExprOption{WantType: &wantType})
@@ -383,11 +361,11 @@ func (r *AwsResourceMissingTagsRule) checkAwsAutoScalingGroupsTags(runner tflint
 	return tags, resourceBlock.DefRange, nil
 }
 
-func (r *AwsResourceMissingTagsRule) emitIssue(runner tflint.Runner, tags map[string]string, config awsResourceTagsRuleConfig, location hcl.Range) {
+func (r *AwsResourceMissingTagsRule) emitIssue(runner tflint.Runner, tags []string, config awsResourceTagsRuleConfig, location hcl.Range) {
 	var missing []string
 	for _, tag := range config.Tags {
-		if _, ok := tags[tag]; !ok {
-			missing = append(missing, fmt.Sprintf("\"%s\"", tag))
+		if !slices.Contains(tags, tag) {
+			missing = append(missing, fmt.Sprintf("%q", tag))
 		}
 	}
 	if len(missing) > 0 {
@@ -405,4 +383,28 @@ func stringInSlice(a string, list []string) bool {
 		}
 	}
 	return false
+}
+
+// getTagsKeysForValue returns a list of known tags from a cty.Value.
+// It returns a boolean indicating whether the keys were known.
+// If _any_ key is unknown, the entire value is considered unknown,
+// since we can't know if a required tag might be matched by the unknown key.
+// Values can be unknown.
+func getKeysForValue(value cty.Value) (keys []string, known bool) {
+	if !value.IsKnown() || value.IsNull() {
+		return nil, false
+	}
+
+	for it := value.ElementIterator(); it.Next(); {
+		k, _ := it.Element()
+
+		// If any key is unknown, return early as any missing tag could be this unknown key.
+		if !k.IsKnown() {
+			return nil, false
+		}
+
+		keys = append(keys, k.AsString())
+	}
+
+	return keys, true
 }

--- a/rules/aws_resource_missing_tags.go
+++ b/rules/aws_resource_missing_tags.go
@@ -390,13 +390,13 @@ func stringInSlice(a string, list []string) bool {
 // If _any_ key is unknown, the entire value is considered unknown, since we can't know if a required tag might be matched by the unknown key.
 // Values are entirely ignored and can be unknown.
 func getKeysForValue(value cty.Value) (keys []string, known bool) {
-	if !value.IsKnown() || value.IsNull() {
+	if !value.CanIterateElements() {
 		return nil, false
 	}
 
 	return keys, !value.ForEachElement(func(key, _ cty.Value) bool {
-		// If any key is unknown, return early as any missing tag could be this unknown key.
-		if !key.IsKnown() {
+		// If any key is unknown or sensitive, return early as any missing tag could be this unknown key.
+		if !key.IsKnown() || key.IsNull() || key.IsMarked() {
 			return true
 		}
 

--- a/rules/aws_resource_missing_tags.go
+++ b/rules/aws_resource_missing_tags.go
@@ -385,26 +385,23 @@ func stringInSlice(a string, list []string) bool {
 	return false
 }
 
-// getTagsKeysForValue returns a list of known tags from a cty.Value.
+// getKeysForValue returns a list of keys from a cty.Value, which is assumed to be a map (or unknown).
 // It returns a boolean indicating whether the keys were known.
-// If _any_ key is unknown, the entire value is considered unknown,
-// since we can't know if a required tag might be matched by the unknown key.
-// Values can be unknown.
+// If _any_ key is unknown, the entire value is considered unknown, since we can't know if a required tag might be matched by the unknown key.
+// Values are entirely ignored and can be unknown.
 func getKeysForValue(value cty.Value) (keys []string, known bool) {
 	if !value.IsKnown() || value.IsNull() {
 		return nil, false
 	}
 
-	for it := value.ElementIterator(); it.Next(); {
-		k, _ := it.Element()
-
+	return keys, !value.ForEachElement(func(key, _ cty.Value) bool {
 		// If any key is unknown, return early as any missing tag could be this unknown key.
-		if !k.IsKnown() {
-			return nil, false
+		if !key.IsKnown() {
+			return true
 		}
 
-		keys = append(keys, k.AsString())
-	}
+		keys = append(keys, key.AsString())
 
-	return keys, true
+		return false
+	})
 }

